connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java

```diff
@@ -884,6 +890,158 @@ private void assertTimeoutException(Runnable operation, String expectedStageDesc
         connect.requestTimeout(DEFAULT_REST_REQUEST_TIMEOUT_MS);
     }
 
+    /**
+     * Tests the logic around enforcement of the
+     * {@link org.apache.kafka.connect.runtime.ConnectorConfig#TASKS_MAX_CONFIG tasks.max}
+     * property and how it can be toggled via the
+     * {@link org.apache.kafka.connect.runtime.ConnectorConfig#TASKS_MAX_ENFORCE_CONFIG tasks.max.enforce}
+     * property, following the test plain laid out in
+     * <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-1004%3A+Enforce+tasks.max+property+in+Kafka+Connect#KIP1004:Enforcetasks.maxpropertyinKafkaConnect-TestPlan">KIP-1004</a>.
+     */
+    @Test
+    public void testTasksMaxEnforcement() throws Exception {
+        String configTopic = "tasks-max-enforcement-configs";
+        workerProps.put(CONFIG_TOPIC_CONFIG, configTopic);
+        connect = connectBuilder.build();
+        // start the clusters
+        connect.start();
+
+        connect.assertions().assertAtLeastNumWorkersAreUp(
+                NUM_WORKERS,
+                "Initial group of workers did not start in time."
+        );
+
+        Map<String, String> connectorProps = defaultSourceConnectorProps(TOPIC_NAME);
+        int maxTasks = 1;
+        connectorProps.put(TASKS_MAX_CONFIG, Integer.toString(maxTasks));
+        int numTasks = 2;
+        connectorProps.put(MonitorableSourceConnector.NUM_TASKS, Integer.toString(numTasks));
+        connect.configureConnector(CONNECTOR_NAME, connectorProps);
+
+        // A connector that generates excessive tasks will be failed with an expected error message
+        connect.assertions().assertConnectorIsFailedAndTasksHaveFailed(
+                CONNECTOR_NAME,
+                0,
+                "connector did not fail in time"
+        );
+
+        String expectedErrorSnippet = String.format(
+                "The connector %s has generated %d tasks, which is greater than %d, "
+                        + "the maximum number of tasks it is configured to create. ",
+                CONNECTOR_NAME,
+                numTasks,
+                maxTasks
+        );
+        String errorMessage = connect.connectorStatus(CONNECTOR_NAME).connector().trace();
+        assertThat(errorMessage, containsString(expectedErrorSnippet));
+
+        // Stop all workers in the cluster
+        connect.workers().forEach(connect::removeWorker);
+
+        // Publish a set of too many task configs to the config topic, to simulate
```

Comment:
I think injecting the config topic while the cluster is offline is the right idea.
WDYT about using a real KafkaConfigBackingStore to do the write instead of duplicating the serialization mechanisms?

Reply:
I considered that approach, but the constructor dependencies for the `KafkaConfigBackingStore` (`WorkerConfigTransformer` (which itself requires a `Worker` instance) and `DistributedConfig` especially) were a bit too cumbersome to instantiate. We could refactor the constructor to provide something a bit friendlier but IMO it's not worth it since the existing manual code is only around 30 lines long.

If you think it'd be worth it I can give it another shot, though.

Reply:
> the constructor dependencies for the KafkaConfigBackingStore (WorkerConfigTransformer (which itself requires a Worker instance)

I think you can get away with a null transformer, the ClusterConfigState has null checks (and shouldn't be instantiated anyway). I agree that creating a full WorkerConfigTransformer is going too far.

> DistributedConfig especially

Is there something I'm not seeing? You already have a workerProps, and it gets filled out with all of the embedded cluster's details during the first cluster start.

Reply:
I'd really rather not rely on the mutability of the `workerProps` field as it makes its way though various stages (builder, instantiating embedded worker, starting embedded worker). I'm also uncomfortable passing in a null transformer--a no-op transformer would be cleaner, but also more verbose.

I'll give it a shot.

Reply:
Okay, trying this again, it's still not looking so great. Even with a null `WorkerConfigTransformer` and relying on the mutability of the `workerProps` field, there's another hurdle in the way: a non-null `ConfigBackingStore.UpdateListener` has to be supplied to the `KafkaConfigBackingStore` instance, or it'll fail when trying to read to the end of the config topic after a call to `putTaskConfigs`.

Implementing that interface, even with no-ops, would nearly double the verbosity of this part of the testing code. We could add some `updateListener != null` checks to the `KafkaConfigBackingStore` class, but IMO it's not worth it to remove those guardrails just to accommodate testing code.

Reply:
Okay thank you for humoring me. I don't think it's a blocker for merge.